Basis bounds safety (#585)
flatironinstitute/nemos: docs/how_to_guide/plot_05_transformer_basis.md

## Setting up the TransformerBasis

At this point, we have an object equipped with the necessary methods. So, all we need to do is concatenate the inputs into a single array and call fit_transform, right? **Not quite.**

> The `TransformerBasis` has two additional requirements: input shape and bounds.

Reviewer: ```suggestion
The `TransformerBasis` has two additional requirements: input shape (all basis objects) and bounds (`Eval` only).
```

or something to make this clear at the beginning of this section

### Setting the `input_shape`

#### The Problem

By default, `TransformerBasis` assumes that `inp` has one column per input in the composition. So in this example, with two one-dimensional basis functions, `TransformerBasis` expects an input of shape `(n_samples, 2)`.

However, our actual input has shape `(n_samples, 6)`:

- `counts` consists of **5 columns**

- `speed` consists of **1 column**

When concatenated, this results in a **6-column input**, and our `TransformerBasis` doesn't know how to split those 6 columns.

> This mismatch leads to an error when calling fit_transform (the cell also sets `bounds`, the second requirement, so that the failure below is about the input shape only; bounds are covered [in the next section](#setting-up-the-bounds)):

Reviewer: ```suggestion
This mismatch leads to an error when calling fit_transform:
```

too much detail I think. the comment in the code block is sufficient.

```

> There is a second requirement: any basis that derives its domain from `bounds` must have `bounds` set before it can be used as a transformer. This is a deliberate safety choice. With `bounds` left unset, the domain is inferred from the data, so during cross-validation each fold would re-fit the domain to its own data range. The transformation would then differ from fold to fold, which is rarely what you want when cross-validating.

Reviewer: I think this content is good, but the order is weird. I'd finish the top-level `Setting up the TransformerBasis` section with something like: "TransformerBasis has two additional requirements: input shape and bounds." and then have separate sub-headers about each of them, each of which has their own sub-headers with something like "The problem", "The solution."

Author: I did reorganize the content, let me know if it matches what you are describing

Let's see this with an example:

```{code-cell} ipython3
:tags: [raises-exception]

# reinstantiate the basis transformer for illustration purposes
composite_basis = counts_basis + speed_basis
trans_bas = (composite_basis).to_transformer()
# set the input shape but NOT the bounds
trans_bas.set_input_shape(counts, speed)
# concatenate the inputs
inp = np.concatenate([counts, speed[:, np.newaxis]], axis=1)
print(inp.shape)

trans_bas.fit_transform(inp)

```

### Define the `bounds`

As shown above, any basis whose domain is defined by `bounds` must have it set before the basis is used as a transformer. Only `Eval` bases carry `bounds`; convolutional bases do not.

Let's set the bounds on the `Eval` component (`speed`) of `composite_basis`.

```{code-cell} ipython3

# get the Eval basis using its label and set the bounds
composite_basis["speed"].bounds = (-2, 2)
```

> 

Reviewer: unrelated to this, but your "Learn more" header is at the wrong level -- it should be `##`, not `###`

```

#### The Solution: Defining the Input Shape

:::{note}

> If each basis accepts a single column of the input, as in [this example](sklearn-how-to), then the default behavior works fine, and `set_input_shape` does not need to be called.

Reviewer: this note should probably be moved to the beginning of the `Setting the input shape` section

```

(transformer-basis-bounds)=
### Setting up the `bounds`

#### The Problem

> Any basis that derives its domain from `bounds` must have `bounds` set before it can be used as a transformer. This is a deliberate safety choice. With `bounds` left unset, the domain is inferred from the data, so during cross-validation each fold would re-fit the domain to its own data range. The transformation would then differ from fold to fold, which is rarely what you want when cross-validating.

Reviewer: ```suggestion
Any basis that derives its domain from `bounds` must have `bounds` set explicitly before it can be used as a transformer. This is a deliberate safety choice. With `bounds` left unset, the domain is inferred from the data, so during cross-validation each fold would re-fit the domain to its own data range. The transformation would then differ from fold to fold, which is rarely what you want when cross-validating.
```

Calling `fit`, `transform`, or `fit_transform` on a `TransformerBasis` whose domain-defining `bounds` are unset raises a `RuntimeError`.

```{code-cell} ipython3
:tags: [raises-exception]

# reinstantiate the basis transformer for illustration purposes
composite_basis = counts_basis + speed_basis
trans_bas = (composite_basis).to_transformer()
# set the input shape but NOT the bounds
trans_bas.set_input_shape(counts, speed)
# concatenate the inputs
inp = np.concatenate([counts, speed[:, np.newaxis]], axis=1)
print(inp.shape)

trans_bas.fit_transform(inp)

```

#### The Solution

> Noting that only `Eval` bases carry `bounds` while convolutional bases do not, we can fully set up the `TransformerBasis` by fixing the bounds for the `Eval` component (`speed`) of `composite_basis`.

Reviewer: ```suggestion
We can finish setting up the `TransformerBasis` by setting the bounds for the `speed` component (`speed`) of `composite_basis`. We do not need to do this for the `position` component because it is `Conv` basis object, and only `Eval` basis objects have `bounds`. 
```

or something like that, just reverse the order of the two points here.
